File: catenary.fs
FeatureScript 2716;
import(path : "onshape/std/common.fs", version : "2716.0");

const SAMPLE_COUNT = 33;
const EPS_LENGTH = 1e-12 * meter;
const EPS_DIRECTION = 1e-12;
const EPS_RATIO = 1e-12;
const MIN_SAG = 1e-5 * meter;
const DOWN = vector(0, -1, 0);

annotation { "Feature Type Name" : "Catenary curve",
             "Manipulator Change Function" : "catenaryManipulatorChange" }
export const catenary = defineFeature(function(context is Context, id is Id, definition is map)
    precondition
    {
        annotation { "Name" : "Start", "Filter" : EntityType.VERTEX, "MaxNumberOfPicks" : 1 }
        definition.start is Query;

        annotation { "Name" : "End", "Filter" : EntityType.VERTEX, "MaxNumberOfPicks" : 1 }
        definition.end is Query;

        annotation { "Name" : "Sag" }
        isLength(definition.sag, NONNEGATIVE_LENGTH_BOUNDS);
    }
    {
        if (isQueryEmpty(context, definition.start))
        {
            throw regenError("Select a start vertex.", ["start"]);
        }
        if (isQueryEmpty(context, definition.end))
        {
            throw regenError("Select an end vertex.", ["end"]);
        }

        var A = evVertexPoint(context, { "vertex" : definition.start });
        var B = evVertexPoint(context, { "vertex" : definition.end });
        var chord = B - A;
        var L = norm(chord);
        if (L < EPS_LENGTH)
        {
            throw regenError("Start and end coincide.");
        }
        var hang = hangDirection(A, B);
        var k = solveK(definition.sag / L);
        var points = [];
        for (var i = 0; i < SAMPLE_COUNT; i += 1)
        {
            var t = i / (SAMPLE_COUNT - 1);
            points = append(points, (1 - t) * A + t * B + definition.sag * profile(t, k) * hang);
        }
        opFitSpline(context, id + "spline", { "points" : points });
        addManipulators(context, id, {
            "sagManipulator" : linearManipulator({
                "base" : (A + B) / 2,
                "direction" : hang,
                "offset" : definition.sag,
                "primaryParameterId" : "sag"
            })
        });
    });

export function catenaryManipulatorChange(context is Context, definition is map, newManipulators is map) returns map
{
    definition.sag = max(newManipulators["sagManipulator"].offset, MIN_SAG);
    return definition;
}

function hangDirection(A is Vector, B is Vector) returns Vector
{
    var u = normalize(B - A);
    var h = DOWN - dot(DOWN, u) * u;
    if (norm(h) < EPS_DIRECTION)
    {
        throw regenError("Chord is vertical: hang direction is undefined.");
    }
    return normalize(h);
}

function coshSagRatio(k is number) returns number
{
    return (cosh(k) - 1) / (2 * k);
}

function solveK(sagRatio is number) returns number
{
    if (sagRatio < EPS_RATIO)
    {
        return 0;
    }
    var hi = 1;
    while (coshSagRatio(hi) < sagRatio && hi < 512)
    {
        hi *= 2;
    }
    if (coshSagRatio(hi) < sagRatio)
    {
        throw regenError("Sag is too large for this chord.");
    }
    var lo = 0;
    for (var i = 0; i < 64; i += 1)
    {
        var mid = (lo + hi) / 2;
        if (coshSagRatio(mid) < sagRatio)
        {
            lo = mid;
        }
        else
        {
            hi = mid;
        }
    }
    return (lo + hi) / 2;
}

function profile(t is number, k is number) returns number
{
    if (k == 0)
    {
        return 4 * t * (1 - t);
    }
    return (cosh(k * (2 * t - 1)) - cosh(k)) / (1 - cosh(k));
}

File: catenary.ts
export interface Vec3 {
  readonly x: number;
  readonly y: number;
  readonly z: number;
}

export interface SampleCatenaryArgs {
  readonly start: Vec3;
  readonly end: Vec3;
  readonly sag: number;
  readonly sampleCount?: number;
}

const DEFAULT_SAMPLE_COUNT = 33;
const EPS_LENGTH = 1e-12;
const EPS_DIRECTION = 1e-12;
const EPS_RATIO = 1e-12;
const DOWN: Vec3 = { x: 0, y: -1, z: 0 };

export function sampleCatenary(args: SampleCatenaryArgs): Vec3[] {
  const n = args.sampleCount ?? DEFAULT_SAMPLE_COUNT;
  if (!Number.isInteger(n) || n < 3) {
    throw new Error("sampleCount must be an integer >= 3");
  }
  if (args.sag <= 0) {
    throw new Error("sag must be positive");
  }
  const chord = sub(args.end, args.start);
  const chordLength = norm(chord);
  if (chordLength < EPS_LENGTH) {
    throw new Error("start and end coincide");
  }
  const hang = hangDirection(args.start, args.end);
  const k = solveK(args.sag / chordLength);
  const points: Vec3[] = [];
  for (let i = 0; i < n; i++) {
    const t = i / (n - 1);
    // (1-t)A + tB is exact at the ends; A + t(B-A) is not.
    const base = lerp(args.start, args.end, t);
    points.push(add(base, scale(hang, args.sag * profile(t, k))));
  }
  return points;
}

function hangDirection(start: Vec3, end: Vec3): Vec3 {
  const chord = sub(end, start);
  const u = scale(chord, 1 / norm(chord));
  const h = sub(DOWN, scale(u, dot(DOWN, u)));
  const hLen = norm(h);
  if (hLen < EPS_DIRECTION) {
    throw new Error("chord is vertical: hang direction is undefined");
  }
  return scale(h, 1 / hLen);
}

function coshSagRatio(k: number): number {
  return (Math.cosh(k) - 1) / (2 * k);
}

function solveK(sagRatio: number): number {
  if (sagRatio < EPS_RATIO) {
    // cosh differences cancel here; 4t(1-t) is the analytic k->0 limit
    return 0;
  }
  let hi = 1;
  while (coshSagRatio(hi) < sagRatio && hi < 512) {
    hi *= 2;
  }
  if (coshSagRatio(hi) < sagRatio) {
    throw new Error("sag too large for chord");
  }
  let lo = 0;
  for (let i = 0; i < 64; i++) {
    const mid = (lo + hi) / 2;
    if (coshSagRatio(mid) < sagRatio) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  return (lo + hi) / 2;
}

function profile(t: number, k: number): number {
  if (k === 0) {
    return 4 * t * (1 - t);
  }
  return (Math.cosh(k * (2 * t - 1)) - Math.cosh(k)) / (1 - Math.cosh(k));
}

function lerp(a: Vec3, b: Vec3, t: number): Vec3 {
  return {
    x: (1 - t) * a.x + t * b.x,
    y: (1 - t) * a.y + t * b.y,
    z: (1 - t) * a.z + t * b.z,
  };
}

function add(a: Vec3, b: Vec3): Vec3 {
  return { x: a.x + b.x, y: a.y + b.y, z: a.z + b.z };
}

function sub(a: Vec3, b: Vec3): Vec3 {
  return { x: a.x - b.x, y: a.y - b.y, z: a.z - b.z };
}

function scale(v: Vec3, s: number): Vec3 {
  return { x: v.x * s, y: v.y * s, z: v.z * s };
}

function dot(a: Vec3, b: Vec3): number {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}

function norm(v: Vec3): number {
  return Math.hypot(v.x, v.y, v.z);
}
